File: higgs/trunk/code/ROS/higgs_autonomous_navigation/mission_manager/src/mission_manager.py
#!/usr/bin/env python

import roslib; roslib.load_manifest('mission_manager')
import rospy
from std_msgs.msg import String
from move_base_msgs.msg import MoveBaseGoal, MoveBaseAction
import actionlib
from actionlib_msgs.msg import GoalStatusArray
import numpy as np
from geometry_msgs.msg import PoseWithCovarianceStamped


import roslib.packages

action_lib_states = ["PENDING", "ACTIVE", "PREEMPTED", "SUCCEEDED","ABORTED","REJECTED",
"PREEMPTING", "RECALLING", "RECALLED", "LOST"]

current_status = 0

first_localization_set = False

def main():
    
    rospy.init_node('mission_manager_node')
    
    rospy.loginfo("Mission manager node started")
    
    
    rospy.loginfo("Mission manager: Loading mission path")
    #Get mission path
    pkg_path = roslib.packages.get_pkg_dir('mission_manager')
    mission_path = np.fromfile(pkg_path+"/missions/patrol_sala_calculo_py.txt", dtype=float, sep=" ")
    mission_path=mission_path.reshape(np.size(mission_path,0)/7,7)
    
    rospy.loginfo("Mission manager: Mission path successfully loaded")

   
    client = actionlib.SimpleActionClient('move_base', MoveBaseAction)
    
    
    #Wait for action server of navigator
    while not client.wait_for_server(rospy.Duration(2.0)) and not rospy.is_shutdown():
        rospy.loginfo("Mission manager: Waiting for the move_base action server to come up")

    rospy.loginfo("Mission manager: move_base action is now available")


    rospy.loginfo("Mission manager: Waiting for initial localization")
    rospy.sleep(rospy.Duration(2.0))
    wait_for_initial_localization()
    rospy.loginfo("Mission manager: Robot has first localization. Ready to send goals")

    #Subscribe to goal status to check on systems goal status
    rospy.Subscriber("/move_base/status", GoalStatusArray, goal_status_callback)
    
    global current_status


    #Sending goals
    for i in range(0, np.size(mission_path,0)):
        print "Sending goal "+str(i+1)+":"
        print str(mission_path[i])
        goal = get_move_base_goal(mission_path, i)
        
        
        client.send_goal(goal)
        result = str(client.wait_for_result())
        
        if current_status != 3:
            rospy.loginfo("Mission manager: will not move for a while because goal is not mine!")
        
        rospy.loginfo("Mission manager: I wait for next goal")
        rospy.sleep(rospy.Duration(5.0))
        rospy.loginfo("Mission manager: Wait ended!")
            
        #Here I will wait for the goal to succeed
        while action_lib_states[current_status] !="SUCCEEDED" and not rospy.is_shutdown():
            print "Will note move to the next goal because state is "+action_lib_states[current_status]
            rospy.sleep(rospy.Duration(0.5))
        
        
        
        if rospy.is_shutdown():
            return

        print "Result "+result
    
    
    #Cancel all goals before exiting
    client.cancel_all_goals()
    
    rospy.loginfo("Mission completed!")

#    raw_input("Press Enter to cancel current goal...")
#    client.cancel_goal()
#
#    raw_input("Press Enter to cancel all goals...")
#    client.cancel_all_goals()


def wait_for_initial_localization():
    global first_localization_set

    rospy.Subscriber("/amcl_pose", PoseWithCovarianceStamped, amcl_pose_callback)

    while not first_localization_set:
        rospy.sleep(0.2)
        
    
        
def amcl_pose_callback(msg):
    global first_localization_set
    
    first_localization_set = True


def goal_status_callback(msg):
    global current_status
    current_status = msg.status_list[0].status
#    rospy.loginfo("Current_status: %s",action_lib_states[current_status]) 

def get_move_base_goal(mission_path, i):
    goal = MoveBaseGoal()    
    goal.target_pose.header.frame_id = "/map"
    goal.target_pose.header.stamp = rospy.Time.now()
    
    goal.target_pose.pose.position.x = mission_path[i][0]
    goal.target_pose.pose.position.y = mission_path[i][1]
    goal.target_pose.pose.position.z = mission_path[i][2]
    goal.target_pose.pose.orientation.w = mission_path[i][3]
    goal.target_pose.pose.orientation.x = mission_path[i][4]
    goal.target_pose.pose.orientation.y = mission_path[i][5]
    goal.target_pose.pose.orientation.z =mission_path[i][6]

    return goal
    
if __name__ == '__main__':

    try:
        main()
    except rospy.ROSInterruptException: pass

File: higgs/trunk/code/ROS/higgs_autonomous_navigation/mission_manager/src/mission_manager.cpp
/*********************************************************************
*
* Author: Carlos Hernandez @ ASLab
*
*	Parameters:
*			- file : for the text file containing the waypoints of the patrol mission
*			-
*	Description:
*	This file implements the mission_manager for a patrol mission in the higgs_autonomous_navigation
*	It is an action client of the move_base ros node: sends goals to it and subscribes to their
*	It result, so a call back is performed to send a new waypoint once the previous is achieved
*
*********************************************************************/

#include <vector>
#include <fstream>
#include <ros/ros.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <actionlib/client/simple_action_client.h>

//#include <tf/transform_datatypes.h>

#include <boost/thread.hpp>

typedef actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> MoveBaseClient;

class MissionManagerNode{

private:
	ros::Rate r;
	MoveBaseClient ac;
	ros::NodeHandle n;
	ros::Subscriber pose_sub;			// accepts new goal positions from rviz

	struct{
	class std::vector<move_base_msgs::MoveBaseGoal> path;  // the path to patrol
	std::vector<move_base_msgs::MoveBaseGoal>::iterator goal_iterator;
	} mission;

    // node parameters
    std::string mission_file;


    // reads the waypoints from a file and loads them in mission.path as a vector of goals
	void loadMission(std::string file){
		std::ifstream infile;
		infile.open(mission_file.c_str(), std::ifstream::in);
		if(!infile.is_open()){
			ROS_ERROR("Mision file not found");
		}

		while(infile.good())	{
			getWaypoint(infile);
		}
		// initialize mission iterator
		mission.goal_iterator = mission.path.begin();
	};

	// read waypoint from file line and append it to the mission.path
	void getWaypoint(std::istream& f){
		move_base_msgs::MoveBaseGoal g;
		g.target_pose.header.frame_id = "map";

		f >> g.target_pose.pose.position.x;
		f >> g.target_pose.pose.position.y;
		f >> g.target_pose.pose.position.z;
		f >> g.target_pose.pose.orientation.w;
		f >> g.target_pose.pose.orientation.x;
		f >> g.target_pose.pose.orientation.y;
		f >> g.target_pose.pose.orientation.z;

		mission.path.push_back(g);

	}

	void sendGoal(move_base_msgs::MoveBaseGoal g){

		//g.target_pose.header.stamp = ros::Time::now();

		// when sending a goal, also subscribe to the actionlib state so as to receive callback when it changes in the server
		ac.sendGoal(g,
		            boost::bind(&MissionManagerNode::done_CB, this, _1, _2)
		            );
	}

	void command_next_point(){

	}



public:
	MissionManagerNode(ros::NodeHandle nh) : ac("move_base", true), r(10) {
		n =nh;
		n.param( "file", mission_file, std::string("/usr/local/ros_aslab/higgs_autonomous_navigation/mission_manager/missions/patrol_sala_calculo_new.txt") );

		pose_sub = n.subscribe<geometry_msgs::PoseStamped>("rviz_goal", 0, boost::bind( &MissionManagerNode::rvizGoal_CB, this, _1 ));

	    ROS_INFO("Waiting for action server to start.");
	    ac.waitForServer();
		loadMission(mission_file);
	    ROS_INFO("Mission loaded.");

		// initialize mission iterator
		mission.goal_iterator = mission.path.begin();


	}

	void mission_loop(){

		// send the first goals, the rest are sent in the callback
		sendGoal(*mission.goal_iterator);

		while( n.ok() ){

		}// while

	}

	// Called once when the goal completes
	void done_CB(const actionlib::SimpleClientGoalState& state,
	              const move_base_msgs::MoveBaseResultConstPtr& result)	{

	    ROS_INFO("Finished in state [%s]", state.toString().c_str());

	    //move to next goal in the path
	    if(mission.goal_iterator++ == mission.path.end())
	    		mission.goal_iterator = mission.path.begin();

	    sendGoal(*mission.goal_iterator);

	}


	void rvizGoal_CB(const geometry_msgs::PoseStamped::ConstPtr& msg){
		ROS_INFO("---> Goal received from rViz: \nPose[ %f %f %f ] \n", msg->pose.position.x, msg->pose.position.y, msg->pose.position.z);
		move_base_msgs::MoveBaseGoal goal;
		goal.target_pose.header = msg->header;
		goal.target_pose.pose = msg->pose;
		sendGoal(goal);
	}


};


int main(int argc, char** argv){
  ros::init(argc, argv, "mission_manager");
  ros::NodeHandle nh;

  MissionManagerNode manager(nh);
	printf("before loop");
  manager.mission_loop();

  return 0;
}
